libraries/Filter/NotchFilter.h

```diff
@@ -29,12 +29,24 @@
 template <class T>
 class NotchFilter {
 public:
+    template <class U> friend class HarmonicNotchFilter;
+
     // set parameters
-    void init(float sample_freq_hz, float center_freq_hz, float bandwidth_hz, float attenuation_dB);
+    virtual void init(float sample_freq_hz, float center_freq_hz, float bandwidth_hz, float attenuation_dB);
+    // For use by DynamicNotchFilter
     T apply(const T &sample);
     void reset();
 
+    // calculate attenuation and quality from provided center frequency and bandwidth
+    static void calculate_A_and_Q(float center_freq_hz, float bandwidth_hz, float attenuation_dB, float& A, float& Q) {
+        const float octaves = log2f(center_freq_hz / (center_freq_hz - bandwidth_hz / 2.0f)) * 2.0f;
```

Comment:
shouldn't we check that (center_freq_hz - bandwidth_hz / 2.0f != 0) ?

Reply:
I've done that in the caller:

// adjust the center frequency to be in the allowable range
    center_freq_hz = constrain_float(center_freq_hz, bandwidth_hz * 0.52f, sample_freq_hz * 0.48f);

This was actually a bug in the original notch, pretty sure you can crash the FC if you select some inappropriate values.